drivers/include/periph/flashpage.h

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C) 2016 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    drivers_periph_flashpage Flash page driver
+ * @ingroup     drivers_periph
+ * @brief       Low-level flash page interface
+ *
+ * @{
+ * @file
+ * @brief       Low-level flash page peripheral driver interface
+ *
+ * This interface provides a very simple and straight forward way for writing
+ * a MCU's internal flash. This interface is only capable of reading, verifying,
+ * and writing complete flash pages, it has no support for partial flash access.
+ * This enables for very slim and efficient implementations.
+ *
+ * A module for more fine-grained access of memory locations can easily be
+ * programmed on top of this interface.
+ *
+ * @note        Flash memory has only a limited amount of erase cycles (mostly
+ *              around 10K times), so using this interface in some kind of loops
+ *              can damage you MCU!
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ */
+
+#ifndef FLASHPAGE_H
+#define FLASHPAGE_H
+
+#include <stdint.h>
+
+#include "periph_cpu.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Per default, we expect the internal flash to start at address 0
+ */
+#ifndef CPU_FLASH_BASE
+#define CPU_FLASH_BASE      (0)
+#endif
+
+/**
+ * @brief   Make sure the page size and the number of pages is defined
+ */
+#ifndef FLASHPAGE_SIZE
+#error "periph/flashpage: FLASHPAGE_SIZE not defined"
+#endif
+#ifndef FLASHPAGE_NUMOF
+#error "periph/flashpage: FLASHPAGE_NUMOF not defined"
+#endif
+
+/**
+ * @brief   Return values used in this interface
+ */
+enum {
+    FLASHPAGE_OK          =  0,     /**< everything succeeded */
+    FLASHPAGE_ERR_NOPAGE  = -1,     /**< invalid page number given */
+    FLASHPAGE_ERR_ERASE   = -2,     /**< error while erasing a page */
+    FLASHPAGE_ERR_WRITE   = -3,     /**< error while writing a page */
+    FLASHPAGE_ERR_NOMATCH = -4      /**< page verification failed, written page
+                                     *   did not match given data */
+};
+
+/**
+ * @brief   Translate the given page number into the page's starting address
+ *
+ * @note    The given page MUST be valid, otherwise the returned address points
+ *          to an undefined memory location!
+ *
+ * @param[in] page      page number to get the address of
+ *
+ * @return              starting memory address of the given page
+ */
+static inline void *flashpage_addr(int page)
+{
+    return (void *)(CPU_FLASH_BASE + (page * FLASHPAGE_SIZE));
```

Comment:
@cgundogan: will ommit the `assert()` here intentionally, as this is checked in the 'not inlined' functions and there is no need to check it everytime here.

@kYc0o: this was solved by our offline discussion today: with the help of #6170 you can do everything you need :-)